pci/src/bus.rs

```diff
@@ -356,8 +356,32 @@ impl BusDevice for PciConfigMmio {
     }
 }
 
+fn shift_and_mask(value: u32, offset: usize, mask: u32) -> usize {
+    ((value >> offset) & mask) as usize
+}
+
+// Parse the MMIO address offset to a (bus, device, function, register) tuple.
+// See https://wiki.osdev.org/PCI_Express#Enhanced_Configuration_Mechanism
+fn parse_mmio_config_address(config_address: u32) -> (usize, usize, usize, usize) {
+    const BUS_NUMBER_OFFSET: usize = 20;
+    const BUS_NUMBER_MASK: u32 = 0x00ff;
+    const DEVICE_NUMBER_OFFSET: usize = 15;
+    const DEVICE_NUMBER_MASK: u32 = 0x1f;
+    const FUNCTION_NUMBER_OFFSET: usize = 12;
+    const FUNCTION_NUMBER_MASK: u32 = 0x07;
+    const REGISTER_NUMBER_OFFSET: usize = 2;
+    const REGISTER_NUMBER_MASK: u32 = 0x3ff;
```

Comment:
Is the value for `REGISTER_NUMBER_MASK` a typo? Shouldn't it be `0x3f`?

Reply:
No, it is not a typo. The length of register segment is 12bit in ECAM. See the link in the comment.

Reply:
register number should not be limited to 6 bit since we want to access the extended configuration space (4096 bytes) .